packages/relay-compiler/language/RelayLanguagePluginInterface.js

```diff
@@ -0,0 +1,246 @@
+/**
+ * Copyright (c) 2013-present, Facebook, Inc.
+ *
+ * This source code is licensed under the MIT license found in the
+ * LICENSE file in the root directory of this source tree.
+ *
+ * @flow
+ * @providesModule RelayLanguagePluginInterface
+ * @format
+ */
+
+'use strict';
+
+// TODO T21875029 ../../relay-runtime/util/RelayConcreteNode
+const RelayConcreteNode = require('RelayConcreteNode');
+
+import type {IRTransform, Root, Fragment} from 'graphql-compiler';
+
+/**
+ * A language plugin allows relay-compiler to both read and write files for any
+ * language.
+ *
+ * When reading, the plugin is expected to parse and return GraphQL tags; and
+ * when writing the plugin is responsible for generating type information about
+ * the GraphQL selections made as well as generating the contents of the
+ * artifact file.
+ *
+ * This interface describes the details relay-compiler requires to be able to
+ * use the plugin. The plugin is expected to have as its main default export a
+ * function that returns an object conforming to this interface.
+ */
+export type PluginInterface = {
```

Comment:
nit: how do you feel about typing the plugin interface as the function that returns this shape itself, rather than just the shape? That way we can also type any arguments to language plugins here, in case they are ever added

Reply:
Good point 👍